params/networkhelper/provider_utils.go

```diff
@@ -0,0 +1,184 @@
+	// Extract embedded providers from the current list
+	embeddedProviders := GetEmbeddedProviders(currentProviders)
+	userProviders := GetUserProviders(newUserProviders)
+
+	// Combine new user providers with the existing embedded providers
+	return append(userProviders, embeddedProviders...)
+}
+
+// ReplaceEmbeddedProviders replaces embedded providers with new ones, retaining user-defined providers.
+func ReplaceEmbeddedProviders(currentProviders, newEmbeddedProviders []params.RpcProvider) []params.RpcProvider {
+	// Extract user-defined providers from the current list
+	userProviders := GetUserProviders(currentProviders)
+	embeddedProviders := GetEmbeddedProviders(newEmbeddedProviders)
+
+	// Combine existing user-defined providers with the new embedded providers
+	return append(userProviders, embeddedProviders...)
+}
+
+// OverrideEmbeddedProxyProviders updates all embedded-proxy providers in the given networks.
+// It sets the `Enabled` flag and configures the `AuthLogin` and `AuthPassword` for each provider.
+func OverrideEmbeddedProxyProviders(networks []params.Network, enabled bool, user, password string) []params.Network {
+	updatedNetworks := make([]params.Network, len(networks))
+	for i, network := range networks {
+		// Deep copy the network to avoid mutating the input slice
+		updatedNetwork := network
+		updatedProviders := make([]params.RpcProvider, len(network.RpcProviders))
+
+		// Update the embedded-proxy providers
+		for j, provider := range network.RpcProviders {
+			if provider.Type == params.EmbeddedProxyProviderType {
+				provider.Enabled = enabled
+				provider.AuthLogin = user
+				provider.AuthPassword = password
+			}
+			updatedProviders[j] = provider
+		}
+
+		updatedNetwork.RpcProviders = updatedProviders
+		updatedNetworks[i] = updatedNetwork
+	}
+
+	return updatedNetworks
+}
+
+func OverrideDirectProvidersAuth(networks []params.Network, authTokens map[string]string) []params.Network {
+	updatedNetworks := make([]params.Network, len(networks))
+	for i, network := range networks {
+		updatedNetwork := network
+		updatedProviders := make([]params.RpcProvider, len(network.RpcProviders))
+
+		for j, provider := range network.RpcProviders {
+			updatedProvider := provider
+
+			if provider.Type == params.EmbeddedDirectProviderType {
+				host, err := extractHost(provider.URL)
+				if err == nil {
+					for suffix, token := range authTokens {
+						if strings.HasSuffix(host, suffix) && token != "" {
+							updatedProvider.AuthType = params.TokenAuth
+							updatedProvider.AuthToken = token
+							break
+						}
+					}
+				}
+			}
+
+			updatedProviders[j] = updatedProvider
+		}
+
+		updatedNetwork.RpcProviders = updatedProviders
+		updatedNetworks[i] = updatedNetwork
+	}
+	return updatedNetworks
+}
+
+func OverrideGanacheToken(networks []params.Network, ganacheURL string, chainID uint64, tokenOverride params.TokenOverride) []params.Network {
```

Comment:
Could you confirm with QA team if this is still being used?

Reply:
I'll remove this.
@antdanchenko confirmed that ganache is no longer in use. Switched to Anvil